### DISCUSSION TITLE
Publish App on Mac OS Store

### DISCUSSION THREAD

**Discussion by ghost** [AvaloniaUI/Avalonia#6326] [Q&A] [answered]
Can we publish the app built using Avalonia on Apple Mac OS Store for distribution? 

Can we target both intel x64 and apple arm64 platforms in a single app for submitting to Mac Store?

How can we add functionality so that if we click any file say image, it opens up our application?

**Comment by maxkatz6** [comment, marked as answer]
All of these questions are not related directly to Avalonia (expect the last one, it's only partially related).
Though, I can try to help you with it.

> Can we publish the app built using Avalonia on Apple Mac OS Store for distribution?

Like any other .NET Core application. The only noticeable difference from Xamarin.Mac, for instance, there is no proper IDE support to generate bundles in VS nor Rider.
I did create macOS bundles in the past, and this topic covers pretty much everything about it - https://docs.avaloniaui.net/docs/distribution-publishing/macos
Although instead of publishing it to the store, my applications were distributed sideloaded. Though I don't see any blockers with it, this topic could help in general (ignoring Xamarin.Mac specific parts). https://docs.microsoft.com/en-us/xamarin/mac/deploy-test/publishing-to-the-app-store/

> Can we target both intel x64 and apple arm64 platforms in a single app for submitting to Mac Store?

IIRC you can't, at least right now. .NET does not support M1 yet. .NET 6 will do, but it's still in preview. 
Also, x64+arm64 in single app was planned to be supported in .NET 6, but it's better to ask .NET team/Microsoft about it. I haven't tried it.
Avalonia does not support .NET 6 M1 yet, but there was successful experiment and it waits for .NET 6 release.

> How can we add functionality so that if we click any file say image, it opens up our application?

Couple of ways. Apart from correct plist manifest configuration (more in apple docs), you need to use macOS API for that:
1. Port your application to Mono and use Xamarin.Mac APIs. It's the worst way in my opinion, because it's very limiting.
2. Use modern .NET Core/.NET 5 with [MonoMac-NetStandard](https://www.nuget.org/packages/MonoMac.NetStandard/). It's the easiest way usually, although it might miss some modern APIs. 
3. Wait for .NET 6 or use it in preview. You will be able to use latest macOS APIs with latest .NET runtime. Probably the best approach in perspective. I haven't tried it yet though.
4. Write a native library to access macOS APIs and use PInvoke to access it from C#. In Avalonia we did it with Avalonia.Native project to not rely on Xamarin.Mac or MonoMac, and make avalonia packages size smaller.

But in your specific case with opening files in Avalonia application, we already did a wrapper API for that https://github.com/AvaloniaUI/Avalonia/pull/5639 . So, you can just subscribe on Application.UrlsOpened to read files/url info. But keep in mind that Avalonia does not/should not cover all macOS APIs, and you will need to use one of these libraries from above to anything not covered by Avalonia or .NET BCL.

**Reply by maxkatz6** [reply]
To summarize it, I believe you should wait for .NET 6 or try to use it in preview.
Although it's possible to do some points from your list right now.

**Reply by ghost** [reply]
Thanks for the detailed explanation. This is really helpful to me.

**Comment by mysteryx93** [comment]
Wow publishing a free app to the Mac Store is incredibly complicated! Requiring both a MacOS virtual machine and a $99 USD / year developer subscription??

For a free app, is there a way to distribute it as a downloadable app that passes the Gatekeeper without paying that yearly subscription?

And any update regarding Mac support in .NET 6?

**Reply by maxkatz6** [reply]
> Wow publishing a free app to the Mac Store is incredibly complicated! Requiring both a MacOS virtual machine and a $99 USD / year developer subscription??

Yes, that's how Apple made it. On top of that app must be sandboxes for the Mac Store.

> For a free app, is there a way to distribute it as a downloadable app that passes the Gatekeeper without paying that yearly subscription?

Kinda. You still need to sign your application with a paid certificate. Developer subscription is only required for notarization (if you don't plan to upload to the store), which is optional as for now.

> And any update regarding Mac support in .NET 6?

Update on what? It works. With .net6 it's a bit easier to package app, but fundamental requirements are still the same.

**Reply by mysteryx93** [reply]
Where do I get that paid certificate and how much is it?

With that, they can download and run the app just fine? But with no OS integration or shortcut whatsoever?

**Reply by mysteryx93** [reply]
This is the information about getting certificates, what part of it needs to be purchased?
https://docs.microsoft.com/en-us/xamarin/mac/deploy-test/publishing-to-the-app-store/certificates-identifiers

Also I see that you must register your Apple Device to your account... will that work with a Virtual Machine, or I actually need to purchase a real Apple computer with a real license?

**Reply by maxkatz6** [reply]
Yeah, Xamarin documentation will help a lot while working on it.
And no, Virtual Machine won't work for signing from what I remember. But you can try.

As I said above, you don't need to buy apple development license to actually build, sign and distribute your app. There won't be problems with gatekeeper, except it will ask user if they trust publisher or not. Notarizing with dev license is needed to avoid that. Or another route - apple store - requires dev license from by definition.

**Reply by maxkatz6** [reply]
Personally, I used macOS hosted on the cloud.
And for local builds and signing I use now personal macmini.

**Reply by mysteryx93** [reply]
alright I'll remove desktop into a client who wants it for his mac

**Reply by maxkatz6** [reply]
To minimize access to the user machine, you at least need to use "codesign" command there. I.e. build a non signed bundle on windows/linux machine, codesign it on mac, and then you can optionally create a dmg package on either mac or linux (windows wsl should also be fine).

**Reply by maxkatz6** [reply]
It seems like MS has found some solution to be able to build and run ios on a real device without mac. I don't know if you can create standalone macos package with this though. 
https://twitter.com/sinclairinat0r/status/1529530487668097024?t=EUilbUxTxk_cMOCyVxGBsg&s=19

**Reply by mysteryx93** [reply]
Requires an Apple Developer Account, which is the $99/year thing.

**Reply by sn4k3** [reply]
Also see: https://github.com/zhlynn/zsign
codesign for linux and windows but I never test it

**Reply by mysteryx93** [reply]
zsign looks interesting; but even the quick road of simple code-signing with non-Apple certificate, the cheapest code-signing certificate is $83/year. Not quite suitable for free apps.

**Reply by sn4k3** [reply]
That is true, more and more is harder to keep free apps going on multi-oses with this kind of policies which does not help the user at all.

**Reply by mysteryx93** [reply]
Perhaps the best option will be to make it a paid app on MacOS, particularly once Avalonia supports iOS. If I want to go down that path, am I required to have a Mac to make it work?

**Reply by sn4k3** [reply]
> am I required to have a Mac to make it work?

It looks like it. zsign was made for iOS, outdated and i dont find any information regarding macOS.
I think you also need to pair a macbook with the certificate. 
Everything is pain with apple...

**Reply by maxkatz6** [reply]
> I required to have a Mac to make it work?

Strictly speaking, you can build everything on linux or window (if you will limit your code with .NET at least).
Only signing requires you to have physical macOS, virtual machine won't work.

**Reply by maxkatz6** [reply]
And yes, that's except if you need m1 builds. For native m1 builds you have to build it on m1 device. Luckly, you can build x64 on the same m1 mac as well, so single mac mini for instance is enough.

**Reply by mysteryx93** [reply]
zsign requires a signing certificate; whether Apple or other, but Apple one is required to import into the store. It doesn't go into how to get that certificate, which is where the Apple Dev Account at $99/year is required. And I believe I read that the Apple Dev Account needs to be paired with a macbook. Still can use a friend's mac via remote control. I'll wait for the next version of Avalonia (or v1) before starting to look at this, that should have better support for iOS and Android. Next question: if I purchase Apple's certificate, can I re-use that one to sign the Android app, or I need to purchase yet another certificate?

**Reply by maxkatz6** [reply]
I don't know if it's possible to reuse certificate. Good question.
I am pretty sure you can export it

**Reply by sn4k3** [reply]
Also for macOS M1 users another solution to run your app will be them to install .NET framework (via cask) and call your app by terminal: dotnet path_to_my.app/Contents/MyApp

That way it will work without any other requirements or sign.
And maybe users can create a fast shortcut to execute that and it wont be a problem.

**Reply by mysteryx93** [reply]
Interesting; but only for Mac's new M1, leaving behind all previous Mac users?

**Reply by sn4k3** [reply]
macOS X64 can still run unsigned packed apps. And then there's no need to do the `dotnet` terminal thing

**Reply by mysteryx93** [reply]
Here's another work-around from [here](https://rixstep.com/2/20191230,00.shtml)

Run `xattr -d com.apple.quarantine [downloaded file]` in a terminal which will nullify that requirement entirely in one sentence!

**Comment by mysteryx93** [comment]
Running unsigned Avalonia app on older 2015 Imac with intel i5, OS MOnterey 12.2.1

Published from Windows

    dotnet publish -r osx-x64 -c Release --self-contained true -p:DebugType=None -p:DebugSymbols=false -p:PublishSingleFile=true -p:PublishTrimmed=true -p:TrimMode=link -o "bin/Publish/osx-x64/"


```
Last login: Sun Jul  3 21:52:29 on ttys000
/Users/jason/Downloads/Player432hz-2.1_Mac_x64/Player432hz ; exit;
jason@Jasons-iMac ~ % /Users/jason/Downloads/Player432hz-2.1_Mac_x64/Player432hz ; exit;
Unhandled exception. System.DllNotFoundException: Unable to load shared library 'libAvaloniaNative' or one of its dependencies. In order to help diagnose loading problems, consider setting the DYLD_PRINT_LIBRARIES environment variable: dlopen(liblibAvaloniaNative, 0x0001): tried: 'liblibAvaloniaNative' (no such file), '/usr/local/lib/liblibAvaloniaNative' (no such file), '/usr/lib/liblibAvaloniaNative' (no such file), '/Users/jason/liblibAvaloniaNative' (no such file), '/usr/local/lib/liblibAvaloniaNative' (no such file), '/usr/lib/liblibAvaloniaNative' (no such file)
   at Avalonia.Native.AvaloniaNativePlatform.CreateAvaloniaNative()
   at Avalonia.Native.AvaloniaNativePlatform.Initialize(AvaloniaNativePlatformOptions options) in /_/src/Avalonia.Native/AvaloniaNativePlatform.cs:line 57
   at Avalonia.AvaloniaNativePlatformExtensions.<>c__DisplayClass0_0`1.<UseAvaloniaNative>b__0() in /_/src/Avalonia.Native/AvaloniaNativePlatformExtensions.cs:line 14
   at Avalonia.Controls.AppBuilderBase`1.Setup() in /_/src/Avalonia.Controls/AppBuilderBase.cs:line 303
   at Avalonia.Controls.AppBuilderBase`1.SetupWithLifetime(IApplicationLifetime lifetime) in /_/src/Avalonia.Controls/AppBuilderBase.cs:line 179
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode) in /_/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 208
   at HanumanInstitute.Player432hz.Program.Main(String[] args)
zsh: abort      /Users/jason/Downloads/Player432hz-2.1_Mac_x64/Player432hz

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Process completed]
```

**Reply by sn4k3** [reply]
remove `-p:PublishTrimmed=true -p:TrimMode=link` from publish and you probally be fine.

**Reply by mysteryx93** [reply]
Still no luck. I deployed it with no trimming and without embedded framework.

Can someone download the 432hz Player and try to run it on MacOS and give the specific instructions? `dotnet` command was saying it requires .NET Framework SDK, not just .NET 6.
https://github.com/mysteryx93/HanumanInstituteApps/releases

**Reply by mysteryx93** [reply]
Running it on MacOS Monterey 12.2.1 (Intel i5) gives

    “Player432hz” cannot be opened because it is from an unidentified developer.

Right-click and Open works... to open the file in a text editor!

`dotnet Player432hz` gives

```
Could not execute because the specified command or file was not found.
Possible reasons for this include:
* You misspelled a built-in dotnet command.
You intended to execute a NET program, but dotnet-Player432hz does not exist.
* You intended to run a global tool, but a dotnet-prefixed executable with this name could not be found on the PATH.
```

**Reply by sn4k3** [reply]
I open your **MacOS_x64.zip** packages, but they dont respect macOS app bundle structure
See here: https://github.com/sn4k3/UVtools/releases/latest download an macOS package and see how content must be.

Also read: https://docs.avaloniaui.net/docs/distribution-publishing/macos

**Reply by maxkatz6** [reply]
> Right-click and Open works... to open the file in a text editor!

You need to open "app" package. It seems you don't have one. 
When you develop app with "net6-macos" target, it will be created automatically by .NET. Otherwise, you need to use [dotnet-bundle](https://github.com/egramtel/dotnet-bundle). Plain executable can't be opened in any normal way on macOS (except `dotnet` command).

> Could not execute because the specified command or file was not found.

Looks like you don't have dotnet installed or available globally. Dotnet runtime should be installed on user's machine and part of the "$PATH" variable to work in this way.

**Reply by mysteryx93** [reply]
Is there a way to set the project TargetFramework dynamically based on target runtime parameter?

**Reply by mysteryx93** [reply]
Added these to my project, alongside `dotnet-bundle`

```
<PropertyGroup>
  <TargetFrameworks>net6.0;net6.0-macos</TargetFrameworks>
  <CFBundleName>Player432hz</CFBundleName> <!-- Also defines .app file name -->
  <CFBundleDisplayName>432hz Player</CFBundleDisplayName>
  <CFBundleIdentifier>com.player432hz</CFBundleIdentifier>
  <CFBundleVersion>2.1</CFBundleVersion>
  <CFBundleShortVersionString>2.1</CFBundleShortVersionString>
  <CFBundlePackageType>APPL</CFBundlePackageType>
  <CFBundleSignature>????</CFBundleSignature>
  <CFBundleExecutable>Player432hz</CFBundleExecutable>
  <CFBundleIconFile>Player432hz.icns</CFBundleIconFile> <!-- Will be copied from output directory -->
  <NSPrincipalClass>NSApplication</NSPrincipalClass>
  <NSHighResolutionCapable>true</NSHighResolutionCapable>

  <!-- Optional -->
  <NSRequiresAquaSystemAppearance>true</NSRequiresAquaSystemAppearance>
</PropertyGroup>
```
```
dotnet msbuild -t:BundleApp -p:RuntimeIdentifier=osx-x64
error MSB4057: The target "BundleApp" does not exist in the project.
```

**Reply by sn4k3** [reply]
Do not complicate, just do a script to create all folder structure and copy files over and zip it...

**Reply by mysteryx93** [reply]
You have a lot more release targets than I do. Why do you have Linux and Arch separate? What is rhel? And do I need to publish arm64 for Mac M1?

**Reply by sn4k3** [reply]
> Why do you have Linux and Arch separate?

Because my project use openCV and i need to compile it for different systems.

> What is rhel? 

Red Hat Enterprise Linux, fedora, centos, etc.

> do I need to publish arm64 for Mac M1?

Yes if you want to provide a native arm64 binnary that runs without emulation. But it require signed or users will have to run your app via dotnet path_to_bin.dll


I have my build script to compile and pack everything in one call.

Non-Windows: https://github.com/sn4k3/UVtools/blob/master/build/createRelease.sh
Windows: https://github.com/sn4k3/UVtools/blob/master/build/createRelease.ps1

On windows i call `createRelease.ps1` it will make the windows release + installer and then will call `createRelease.sh` per runtime and do the stuff on a WSL
You may look and adapt for your app if you want. If your app dont require nothing fancy you can use github actions to do the release for you.

**Reply by mysteryx93** [reply]
I was about to ask you for such script, as it feels like re-inventing the wheel. One complication is that while I deploy AppImage using a Linux bash script, Windows binaries needs to be compiled in Windows to get icons and attributes... so although the project is cross-platform, the build is platform-specific and requires both OS!

**Reply by sn4k3** [reply]
I do everything in my windows host with WSL
As told i run [createRelease.ps1](https://github.com/sn4k3/UVtools/blob/master/build/createRelease.ps1) once and it create the all packages you see in my release.

While i use WSL you can create all releases under windows (with execption of AppImage), i just have the [createRelease.sh](https://github.com/sn4k3/UVtools/blob/master/build/createRelease.sh) for linux and macos guys can also create releases of the project without the need of a windows machine. So the powershell calls the wsl sh script

**Reply by mysteryx93** [reply]
Got it working! Here are my scripts. In my case I got several projects in the solution to publish. I created `publish-osx App [runtime]`, and publish-linux and publish-win with the same syntax. Then publish-osx-all, publish-linux-all and publish-win-all that compiles all projects in one swoop.

https://github.com/mysteryx93/HanumanInstituteApps/tree/master/Build

**Comment by mysteryx93** [comment]
Can someone confirm whether this method of bypassing the Gatekeeper works too on all MacOS versions?
https://support.apple.com/en-us/HT202491
